### PR TITLE
Migrated to jakarta.servlet package and Spring 6.

### PR DESCRIPTION
**Pull request by allantodd**
Migrated to jakarta.servlet package and Spring 6.  Deleted I18nStruts2Adapter.  Deleted tests related to portlets.

Note that test fail - httpunit does not support jakarta.servlet namespace

### CODE REVIEW

**Comment by hazendaz** [comment]
@allantodd Please bump httpunit to 2.0.0 to see if that works out as solving jakarta for that module.

File: pom.xml [hunk 1]
```diff
@@ -295,7 +295,7 @@
           <dependency>
             <groupId>net.trajano.wagon</groupId>
             <artifactId>wagon-git</artifactId>
-            <version>2.0.3</version>
+            <version>2.0.4</version>
           </dependency>
```
Comment (hazendaz) [inline pom.xml:299]: @allantodd move wagon-git back to 2.0.3.  The release will fail if 2.0.4 for 2 reasons.  1 it was coded wrong and 2 its defunct and won't be fixed.  Only 2.0.3 is valid.

File: displaytag/pom.xml [hunk 2]
```diff
@@ -20,7 +20,7 @@
   </scm>
 
   <properties>
-    <tomcat.version>7.0.109</tomcat.version>
+    <tomcat.version>11.0.0-M1</tomcat.version>
```
Comment (hazendaz) [inline displaytag/pom.xml:23]: @allantodd Make this 10.1.x series not 11 milestone.

File: displaytag/pom.xml [hunk 3]
```diff
@@ -198,7 +198,7 @@
     <dependency>
       <groupId>net.sourceforge.htmlunit</groupId>
       <artifactId>neko-htmlunit</artifactId>
-      <version>2.67.0</version>
+      <version>2.70.0</version>
```
Comment (hazendaz) [inline displaytag/pom.xml:201]: @allantodd Won't work past 2.67.0, please go back to that.  Only way newer will work is figuring out 1 class in httpunit that I've been unsuccessful figuring out.  htmlunit project there internalized a lot of xerces so things that did work don't now.

File: displaytag/pom.xml [hunk 4]
```diff
@@ -289,6 +289,13 @@
       <version>3.0.4</version>
       <scope>test</scope>
     </dependency>
+    <!-- just keep the tests compiling - they don't work though -->
```
Comment (hazendaz) [inline displaytag/pom.xml:292]: @allantodd remove this one.

File: pom.xml (hunk 1 above)
Comment (allantodd) [reply pom.xml]: done

File: displaytag/pom.xml (hunk 3 above)
Comment (allantodd) [reply displaytag/pom.xml]: done

File: displaytag/pom.xml (hunk 4 above)
Comment (allantodd) [reply displaytag/pom.xml]: done

File: displaytag/pom.xml (hunk 2 above)
Comment (allantodd) [reply displaytag/pom.xml]: done

**Comment by allantodd** [comment]
@hazendaz thanks.  I have made the changes you requested. The tests still fail but it looks like progress.  The failures appear to be all the same problem (though I haven't checked every one)

`java.lang.NullPointerException: Cannot invoke "org.apache.jasper.compiler.TldCache.getTldResourcePath(String)" because the return value of "org.apache.jasper.Options.getTldCache()" is null`

for example

```
[ERROR] org.displaytag.properties.TitleKeyAutoColumnSpringTest.doTest  Time elapsed: 0.686 s  <<< ERROR!
com.meterware.httpunit.HttpInternalErrorException: Error on HTTP request: 500 org.apache.jasper.JasperException: org.apache.jasper.JasperException: Error processing file [/jsps/titlekeyautocolumn.jsp] [http://localhost/context/jsps/titlekeyautocolumn.jsp.spring]
        at com.meterware.servletunit.ServletUnitClient.newResponse(ServletUnitClient.java:142)
        at com.meterware.httpunit.WebClient.createResponse(WebClient.java:648)
        ...
        ...
Caused by: org.apache.jasper.JasperException: org.apache.jasper.JasperException: Error processing file [/jsps/titlekeyautocolumn.jsp]
        at org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:578)
        at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:422)
        at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:380)
        at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:328)
        ...
        ...
Caused by: org.apache.jasper.JasperException: Error processing file [/jsps/titlekeyautocolumn.jsp]
        at org.apache.jasper.compiler.DefaultErrorHandler.jspError(DefaultErrorHandler.java:54)
        at org.apache.jasper.compiler.ErrorDispatcher.dispatch(ErrorDispatcher.java:294)
        at org.apache.jasper.compiler.ErrorDispatcher.jspError(ErrorDispatcher.java:142)
        ...
        ...
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.jasper.compiler.TldCache.getTldResourcePath(String)" because the return value of "org.apache.jasper.Options.getTldCache()" is null
        at org.apache.jasper.JspCompilationContext.getTldResourcePath(JspCompilationContext.java:569)
        at org.apache.jasper.compiler.JspDocumentParser.getTaglibInfo(JspDocumentParser.java:1309)
        at org.apache.jasper.compiler.JspDocumentParser.startPrefixMapping(JspDocumentParser.java:831)
        ...
        ...

```

**Comment by allantodd** [comment]
I found some remaining portlet code and a dependency so removed it on another commit.  There seems to be no migration path for portlets to use the jakarta packages.

Confirmed that the code still works for me

**Comment by hazendaz** [comment]
@allantodd Please merge in master now and resolve couple merge conflicts.  This will now also run on GH-actions.  

One questino: Why the removal of I18nStruts2Adaptor?  That is still valid (struts 2).  Or is that having issues?

**Comment by hazendaz** [comment]
@allantodd Sorry tests should also start working better, repo needed the work for tomcat 8+ i had not pushed up previously.  Its on master now.  I had to disable 9 tests as a result that I could not otherwise resolve.

**Comment by allantodd** [comment]
Thanks @hazendaz, that's compiling and the tests run okay now.  I'll test it out later and push.

I had to remove `I18nStruts2Adaptor` because the latest Struts version still uses the `javax.servlet` namespace.

This compile error occurs (after changing the imports to use jakarta namespace)

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project displaytag-jakarta: Compilation failure
[ERROR] /Users/allantodd/projects/displaytag/displaytag/src/main/java/org/displaytag/localization/I18nStruts2Adapter.java:[121,42] cannot access javax.servlet.jsp.PageContext
[ERROR]   class file for javax.servlet.jsp.PageContext not found
```

The line in question is

`final ValueStack stack = TagUtils.getStack(pageContext);`

The error occurs because Struts' `TagUtils` class imports from `javax.servlet` - see https://github.com/apache/struts/blob/master/core/src/main/java/org/apache/struts2/views/jsp/TagUtils.java

The specification versions that Struts 6.1.1 supports are listed at https://struts.apache.org/announce-2022#a20221128

**Comment by allantodd** [comment]
Hi @hazendaz I've rebased onto your latest master and squashed.  The tests are passing and the library is working for me

**Comment by hazendaz** [comment]
> Thanks @hazendaz, that's compiling and the tests run okay now. I'll test it out later and push.
> 
> I had to remove `I18nStruts2Adaptor` because the latest Struts version still uses the `javax.servlet` namespace.
> 
> This compile error occurs (after changing the imports to use jakarta namespace)
> 
> ```
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project displaytag-jakarta: Compilation failure
> [ERROR] /Users/allantodd/projects/displaytag/displaytag/src/main/java/org/displaytag/localization/I18nStruts2Adapter.java:[121,42] cannot access javax.servlet.jsp.PageContext
> [ERROR]   class file for javax.servlet.jsp.PageContext not found
> ```
> 
> The line in question is
> 
> `final ValueStack stack = TagUtils.getStack(pageContext);`
> 
> The error occurs because Struts' `TagUtils` class imports from `javax.servlet` - see https://github.com/apache/struts/blob/master/core/src/main/java/org/apache/struts2/views/jsp/TagUtils.java
> 
> The specification versions that Struts 6.1.1 supports are listed at https://struts.apache.org/announce-2022#a20221128

ok lets do this, please make 2 commits here, one for the upgrade and second for removal of struts.  I looked up and struts is going to do the update in 7.0.0 but waiting on commons fileupload and probably others before they can.  So if you make 2 commits here, that makes it easy to add back later.  Please then open a separate issue here that we need to add back and reference the commit.  I'll add more info to that.

Once that is done, will merge and then get this released properly.  Thanks.

**Comment by allantodd** [comment]
Thanks @hazendaz , I've split into two commits.  I removed Struts in the first commit, before the servlet spec upgrades, to avoid having a commit that doesn't compile.  I trust this is okay but I can reverse it if you prefer.

**Comment by hazendaz** [comment]
I'll cleanup CI builds after here and get release as soon as I can, probably in few days.  Somewhere in here obviously we didn't care to split out the portlet code as that is long dead.  The last release should be good enough there and spring long deleted hence all their code I copied from earlier spring.

**Comment by hazendaz** [comment]
@allantodd Thanks for putting up with me on this one 👍 Glad to see this move to jakarta now.  Renovate will get a few more updates and I'll double check as some portions of EE moved to newly named items...thinking if we have EL impl its expressly now, or mail is angus, etc.  I'll take care of those parts as needed before release.  Appreciate your work here...

**Comment by allantodd** [comment]
@hazendaz, you're welcome - it's been good cooperating on this.  Thanks for all your work maintaining this library.  One suggestion (sorry if I'm stating the obvious here) - might be worth updating your major version to e.g. 3.0.0 as these are breaking changes.

**Comment by hazendaz** [comment]
Yes will bump to 3.0.0 once released.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Pull request merged**